### ISSUE TITLE
DismissEffect does not work on iOS14

### ISSUE THREAD

**Issue by sk409** [pointfreeco/swift-composable-architecture#2233]
### Description

DismissEffect does not work on iOS14.

Nothing happens when I run the code below on iOS14.

```swift
struct CoreReducer: ReducerProtocol {
    @Dependency(\.dismiss) private var dismiss

    func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
        switch action {
        ...
            return .fireAndForget {
                // Nothing happens.
                await dismiss()
            }
        }
    }
}
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Screen closes after executing DismissEffect.

### Actual behavior

Screen does not close after executing DismissEffect.

### Steps to reproduce

1. Transition to any screen on iOS14
2. Execute DismissEffect on the transition destination screen

### The Composable Architecture version information

0.54.1

### Destination operating system

iOS 14.5

### Xcode version information

14.3

### Swift Compiler version information

```shell
swift-driver version: 1.75.2 Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)
```

**Comment by stephencelis** [comment]
@sk409 Can you attach a sample project reproducing the problem? The issue doesn't have enough to go by.

Note that the `dismiss` effect only works with features embedded with `ReducerProtocol.ifLet`, held in `PresentationState`/`PresentationAction`, and presented via an appropriate view helper, like `View.sheet(store:)`.

**Comment by sk409** [comment]
@stephencelis 

This is the sample project to reproduce.

https://github.com/sk409/DissmissEffectSample

The code using the DismissEffect is below.

https://github.com/sk409/DissmissEffectSample/blob/main/Test/ContentView.swift

**Comment by sk409** [comment]
@stephencelis 

> Note that the dismiss effect only works with features embedded with ReducerProtocol.ifLet, held in PresentationState/PresentationAction, and presented via an appropriate view helper, like View.sheet(store:).

I'm using DismissEffect in the correct order, so it's working on iOS15 and iOS16.

**Comment by stephencelis** [comment]
@sk409 Yup I took a look at your code but need to download the iOS 14 runtime sometime later today to investigate. It may just have to do with the fact that programmatic navigation in SwiftUI is quite buggy. It's better in iOS 16.4 and 17, but previous iOSes do not perform well when it comes to navigation.

**Comment by stephencelis** [comment]
Thanks again for the repro! Finally had a chance to sit down with this today and have a fix here: #2283. Please confirm that it fixes the problem for you!

**Comment by sk409** [comment]
@stephencelis 
I was able to confirm that DismissEffect is working on iOS14.
Thank you for your response.
